### ISSUE TITLE
Make compiled models serializable

### ISSUE THREAD

**Issue by mariosasko** [pytorch/pytorch#101107]
### 🐛 Describe the bug

Serializing a compiled model with `pickle` fails with `Can't pickle local object 'convert_frame.<locals>._convert_frame'` and `cannot pickle 'ConfigModuleInstance' object` when using `dill`.

A Colab with an example: 
https://colab.research.google.com/drive/1v6jUUq86ql1Era4X47cIDj7bzrrz2RZe?usp=sharing

In Hugging Face Datasets, this error stops us from generating (deterministic) hashes for transforms (functions) that reference a compiled model, meaning such transforms cannot be cached and must be re-computed each time when transforming a dataset.   

(The "export" API for the compiled models would also work for us.)

### Versions

<details>
<summary>Colab env with torch 2.0.1 installed</summary>

```
PyTorch version: 2.0.1+cu118
Is debug build: False
CUDA used to build PyTorch: 11.8
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.5 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
Clang version: 10.0.0-4ubuntu1 
CMake version: version 3.25.2
Libc version: glibc-2.31

Python version: 3.10.11 (main, Apr  5 2023, 14:15:10) [GCC 9.4.0] (64-bit runtime)
Python platform: Linux-5.10.147+-x86_64-with-glibc2.31
Is CUDA available: False
CUDA runtime version: 11.8.89
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: Could not collect
Nvidia driver version: Could not collect
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.7.0
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Byte Order:                      Little Endian
Address sizes:                   46 bits physical, 48 bits virtual
CPU(s):                          2
On-line CPU(s) list:             0,1
Thread(s) per core:              2
Core(s) per socket:              1
Socket(s):                       1
NUMA node(s):                    1
Vendor ID:                       GenuineIntel
CPU family:                      6
Model:                           79
Model name:                      Intel(R) Xeon(R) CPU @ 2.20GHz
Stepping:                        0
CPU MHz:                         2200.196
BogoMIPS:                        4400.39
Hypervisor vendor:               KVM
Virtualization type:             full
L1d cache:                       32 KiB
L1i cache:                       32 KiB
L2 cache:                        256 KiB
L3 cache:                        55 MiB
NUMA node0 CPU(s):               0,1
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Mitigation; PTE Inversion
Vulnerability Mds:               Vulnerable; SMT Host state unknown
Vulnerability Meltdown:          Vulnerable
Vulnerability Mmio stale data:   Vulnerable
Vulnerability Retbleed:          Vulnerable
Vulnerability Spec store bypass: Vulnerable
Vulnerability Spectre v1:        Vulnerable: __user pointer sanitization and usercopy barriers only; no swapgs barriers
Vulnerability Spectre v2:        Vulnerable, IBPB: disabled, STIBP: disabled, PBRSB-eIBRS: Not affected
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Vulnerable
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single ssbd ibrs ibpb stibp fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm rdseed adx smap xsaveopt arat md_clear arch_capabilities

Versions of relevant libraries:
[pip3] numpy==1.22.4
[pip3] torch==2.0.1+cu118
[pip3] torchaudio==2.0.2+cu118
[pip3] torchdata==0.6.0
[pip3] torchsummary==1.5.1
[pip3] torchtext==0.15.1
[pip3] torchvision==0.15.2+cu118
[pip3] triton==2.0.0
[conda] Could not collect
```
</details>

**Comment by msaroufim** [comment]
This has come a few times so aggregating how this works today and our plans for next steps in the future
1. Indeed you can't pickle an optimized module but you can pickle the original module because the weights are shared https://pytorch.org/get-started/pytorch-2.0/#serialization
2. Because 1 is annoying we recently introduced an in place module compilation API that would make saving and loading work https://github.com/pytorch/pytorch/pull/97565
3. But unfortunately 2 doesn't solve the problem of having to recompile a model when you load it so cold starts for inference are bad, I'll have a POC working to solve this very soon
4. Export obviously is a good solution but so far no date for an official release, export is also focused on environments without python available but if python is available I think 3 will work just fine

**Comment by msaroufim** [comment]
@mariosasko just wanna make sure I understand, as a first step it sounds like you're mostly interested in knowing exactly which args were using when compiling a model for reproducibility. If so I was also planning on just putting that in the nn module state dict

**Comment by varunshenoy** [comment]
Any updates on this, or has anyone found a workaround?

**Comment by msaroufim** [comment]
So the simplest workaround is to save the state dict and not the model - I tired to get this to work here https://github.com/pytorch/pytorch/pull/101651 and it did work you could effectively save compiled models directly but the problem was my changes introduced some extra graph breaks across the board which have a performance impact, I couldn't figure it out and I don't have bandwidth to further inspect but if someone would like to revisit I'd be happy to review and merge

**Comment by YodaEmbedding** [comment]
My workaround for "repairing" checkpoints that contain the undesired `"_orig_mod."` prefix:

```python
import sys
import torch


def remove_prefix(text, prefix):
    if text.startswith(prefix):
        return text[len(prefix) :]
    return text


def repair_state_dict(in_state_dict, prefix="_orig_mod."):
    out_state_dict = {}
    for src_key, value in in_state_dict.items():
        if src_key.startswith(prefix):
            dest_key = remove_prefix(src_key, prefix)
        else:
            dest_key = src_key
        print(f"{src_key}  ==>  {dest_key}")
        out_state_dict[dest_key] = value
    assert len(out_state_dict) == len(in_state_dict)  # Ensure one-to-one mapping.
    return out_state_dict


def repair_checkpoint(path):
    ckpt = torch.load(path)
    in_state_dict = ckpt
    out_state_dict = repair_state_dict(in_state_dict)
    if set(in_state_dict.keys()) == set(out_state_dict.keys()):
        return  # Do not write checkpoint if no need to repair!
    ckpt = out_state_dict
    torch.save(ckpt, path)


paths = sys.argv[1:]
for path in paths:
    print(path)
    repair_checkpoint(path)
    print("========")
```

Then:

```bash
python checkpoint_unwrap_orig_model.py **/*.pth.tar
```

**Comment by wilson97** [comment]
@msaroufim I tried your workaround (torch.compile my model, then save the state_dict, then load a new non-compiled version, insert the saved state_dict) and the model is slow (non-compiled). Am I doing something wrong? Or are you saying that I have to recompile regardless of whether I load a state_dict or not?

**Comment by pallgeuer** [comment]
> My workaround is to "repair" checkpoints that contain the undesired `"_orig_mod."` prefix.
> 
> Save the following script:
> 
> ```python
> import sys
> import torch
> 
> 
> def remove_prefix(text, prefix):
>     if text.startswith(prefix):
>         return text[len(prefix) :]
>     return text
> 
> 
> def repair_checkpoint(path):
>     ckpt = torch.load(path)
>     in_state_dict = ckpt["model_state_dict"]
>     pairings = [
>         (src_key, remove_prefix(src_key, "_orig_mod."))
>         for src_key in in_state_dict.keys()
>     ]
>     if all(src_key == dest_key for src_key, dest_key in pairings):
>         return  # Do not write checkpoint if no need to repair!
>     out_state_dict = {}
>     for src_key, dest_key in pairings:
>         print(f"{src_key}  ==>  {dest_key}")
>         out_state_dict[dest_key] = in_state_dict[src_key]
>     ckpt["model_state_dict"] = out_state_dict
>     torch.save(ckpt, path)
> 
> 
> if __name__ == "__main__":
>     paths = sys.argv[1:]
>     for path in paths:
>         print(path)
>         repair_checkpoint(path)
>         print("========")
> ```
> 
> Then:
> 
> ```shell
> python checkpoint_unwrap_orig_model.py **/*.pth
> ```
> 
> NOTE: In my checkpoints, the state_dict is actually inside `ckpt["model_state_dict"]`. If yours is in a different place, adjust that as necessary, e.g. `ckpt` if your state_dict is exactly the root of the checkpoint.

If possible (i.e. because you can change the code that is generating the checkpoints), you can save `model._orig_mod.state_dict()` instead of `model.state_dict()` for compiled models. This will avoid the `_orig_mod.` prefix everywhere, and therefore not require and fixing on load. You just need to `load_state_dict()` and `torch.compile()` after loading.

**Comment by fxmarty** [comment]
This is a duplicate of https://github.com/pytorch/pytorch/issues/93470

**Comment by ecstayalive** [comment]
I usually override the state_dict() method and load_state_dict() method, using a specific structure to solve most problems, for example:
`
from collections import OrderedDict

class CustomNet(nn.Module):
    def __init__(self):
        self.kernel: nn.Module = ...
    # override state_dict() method
    def state_dict(self, *args, destination=None, prefix="", keep_vars=False):
        if destination is None:
            destination = OrderedDict()
        prefix = ""  # remove prefix
        destination.update([('kernel', self.kernel.state_dict(destination=destination, prefix=prefix, keep_vars=keep_vars)])
        return destination
    # override load_state_dict()
    def load_state_dict(self, state_dict, ...):
        self.kernel.load_state_dict(state_dict)
`
This structure can fix '_orig_mod' prefix. However,  the above code is just an example of the method and cannot be run. You should adjust the structure accordingly according to your own code.

**Comment by tugsbayasgalan** [comment]
@msaroufim any updates?

**Comment by msaroufim** [comment]
Nope unsassigning myself for now since i haven't had time to keep fixing issues here

**Comment by fxmarty** [comment]
Isn't `torch.export.export` kind of similar to what is requested here? https://pytorch.org/docs/stable/export.html

**Comment by yf225** [comment]
It feels like this should be tackled by export. Assigning `oncall: export` label but feel free to revert it.

**Comment by rfeinman** [comment]
@fxmarty `torch.export.export` only compiles a single path from the program control flow, as far as I understand.

It would be great to be able to save a model that has run `torch.compile` so that we do not need to re-compile each time we launch a program! +1 for this

**Comment by angelayi** [comment]
@rfeinman Could you clarify what you mean by "torch.export.export only compiles a single path from the program control flow"? Export should be able to handle control flow if it is rewritten using [`torch.cond`](https://pytorch.org/docs/main/cond.html).

**Comment by rfeinman** [comment]
@angelayi my understanding is based off of the explanation here: https://pytorch.org/tutorials/intermediate/torch_compile_tutorial.html#comparison-to-torchscript-and-fx-tracing

> Primarily, the advantage of torch.compile lies in its ability to handle arbitrary Python code with minimal changes to existing code.

> One case that torch.compile can handle that other compiler solutions struggle with is data-dependent control flow (the if x.sum() < 0: line below).

> TorchScript tracing f1 results in silently incorrect results, since only the actual control flow path is traced.

I think the ability of `torch.compile` to handle arbitrary Python code with minimal changes is a very nice feature, and it would be great if this feature could transfer to serialization (i.e., if we don't have to swap in `torch.cond`, etc).

**Comment by angelayi** [comment]
@rfeinman that makes sense! torch.export wants to get a full graph representation of the code so it requires these code rewrites, instead of defaulting to the python code, which is what torch.compile does.

**Comment by fxmarty** [comment]
Maybe a single GraphModule is able to represent torch.cond then? I was not aware of that

**Comment by vgoklani** [comment]
Curious has this issue been resolved?

Essentially the goal is to saved a compiled model, so that we don't have to re-compile each time. (i.e. serialize the compiled model).

**Comment by fernandoFernandeSantos** [comment]
I'm curious as well, this would be a very nice feature to have

**Comment by ezyang** [comment]
cc @zhxchen17

**Comment by j-adamczyk** [comment]
+1, this would be very useful. Right now, I have to recompile the model on each Docker container startup. This is especially bad when dynamically scaling up Kubernetes microservices or with Lambda functions, when a requests surge occurs. Particularly since I use very static embeddings models (e.g. CLIP, DINOv2), with inputs scaled to the same size, so serializing those models after compilation should be relatively easy.

**Comment by vgoklani** [comment]
I've tried using the new `artifacts`

it does seem faster, but still takes a while



```python
    artifacts = torch.compiler.save_cache_artifacts()

    assert artifacts is not None
    artifact_bytes, cache_info = artifacts

    with gzip.open(artifacts_filename, "wb") as f:
        f.write(artifact_bytes)
        log.info(f"successfully saved artifacts: {cache_info}")
```

```python
        with gzip.open(artifacts_filename, "rb") as f:
            artifact_bytes = f.read()

        cache_info = torch.compiler.load_cache_artifacts(artifact_bytes)
```

**Comment by zhxchen17** [comment]
@vgoklani Yeah probably because load_cache_artifacts() cache mostly aotautograd and inductor artifacts.

Just a small update that we are developing a different frontend which will further strip the compilation artifacts from dynamo which hopefully will reduce the loading time to near zero.

**Comment by ffox9527** [comment]
Has this problem been solved now?

**Comment by msaroufim** [comment]
There have been significant updates made to https://docs.pytorch.org/tutorials/recipes/torch_compile_caching_tutorial.html

**Comment by fernandoFernandeSantos** [comment]
Thanks for the points @msaroufim 
The docs say that 

> It is important to note that caching validates that the cache artifacts are used with the same PyTorch and Triton version, as well as, same GPU when device is set to be cuda.

If I know that my setup will be static (Pytorch, Triton, and CUDA drivers will not change), the ideal scenario would be to disable those verifications. I don't know if it can be done with caching

**Comment by ffox9527** [comment]
> There have been significant updates made to https://docs.pytorch.org/tutorials/recipes/torch_compile_caching_tutorial.html

I am not very familiar with this update. Will this solve our problem?
